plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Core/Zip.hs

```diff
@@ -0,0 +1,70 @@
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeFamilies        #-}
+module UntypedPlutusCore.Core.Zip
+    ( pzipWith
+    , pzip
+    , tzipWith
+    , tzip
+    ) where
+
+import Control.Monad (when)
+import PlutusPrelude
+import UntypedPlutusCore.Core.Instance.Eq ()
+import UntypedPlutusCore.Core.Type
+
+-- | Zip two programs using a combinator function for annotations.
+-- Fail if the input programs are not "equal" modulo annotations.
+pzipWith :: forall p name uni fun ann1 ann2 ann3 m.
+           (p ~ Program name uni fun, (Eq (Term name uni fun ())), MonadFail m)
+         => (ann1 -> ann2 -> ann3)
+         -> p ann1
+         -> p ann2
+         -> m (p ann3)
+pzipWith f (Program ann1 ver1 t1) (Program ann2 ver2 t2) = do
+    when (void ver1 /= void ver2) $
+       fail "Versions do not match."
+    let ver1Combined = over verAnn (`f` (ver2^.verAnn)) ver1
+    Program (f ann1 ann2) ver1Combined <$> tzipWith f t1 t2
+
+-- | Zip two terms using a combinator function for annotations.
+-- Fail if the input terms are not "equal" modulo annotations.
+-- TODO: this is not an optimal implementation
+tzipWith :: forall t name uni fun ann1 ann2 ann3 m.
+           (t ~ Term name uni fun, Eq (t ()), MonadFail m)
+         => (ann1 -> ann2 -> ann3)
+         -> t ann1
+         -> t ann2
+         -> m (t ann3)
+tzipWith f term1 term2 = do
+    -- first establish that terms are equal modulo annotations
+    -- slower this way but avoids some equality boilerplate for now
+    when (void term1 /= void term2) $
+       fail "Terms do not match."
```

Comment:
Why does this help? Doesn't the catch-all pattern match below account for this regardless?

Reply:
I thnk there are some checks missing from this zip function, right? 
Checks missing:
- Alpha equivalence check in case t is `Term Name`.
- Eq uni and Eq fun check for builtins.

So I thought to stay on the safe side.

Btw, I should add a comment f input terms t1 and t2 is `Term Name` , the output is "left-biased" and will return the names of t1.

Reply:
I see. I'd be tempted to do an equality check on the builtins/constants/names, but I gues this is also okay? But write something down!